cluster/ci/origin/logging-infra/gcs.go

```diff
@@ -0,0 +1,82 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"sync"
+
+	"cloud.google.com/go/storage"
+)
+
+// Metadata holds run-time data about the build
+type Metadata struct {
+	// JobName is the name of the job triggering
+	// this build
+	JobName string `json:"job-name"`
+
+	// BuildNumber is the identifier for this build
+	// of the job
+	BuildNumber int `json:"build-number"`
+
+	// Parameters contains all parameters passed
+	// to this build
+	Parameters map[string]string `json:"parameters"`
+}
+
+// GcsPath determines where artifacts for this job should
+// be uploaded to under the base bucket
+func (m *Metadata) GcsPath() string {
```

Comment:
@csrwng @smarterclayton this is where I am a little unclear of what the best course of action is. Right now, we seem to be in a place where a log needs to be at most in two places, and one of those places is infinitely more common as a destination for browsing, so we place the log there (`pr-logs/pull/{pull}/{name}/{build}/`) and do the GCS "symlink" hack to also place a copy in the other location (`pr-logs/directory/{name}/{build}/`).

As we move to having one build step that creates the image that N tests end up running on, we no longer have two places and no longer is any one place more favored. I don't see a good way to preserve backwards compatibility unless today we have this logic here upload the literal file to every possible place it will be needed, which only would work for the first set of jobs anyway. If new jobs are triggered under the same commit, we'd need to copy the logs into their directories too so they would link to the original build log.

I'm not really sure if this is a solvable problem for Gubernator or not. I would love to use a simple schema where build logs are indexed by job name and build number like in this function, and a RDBMS sits in front, reads all this metadata, and allows for the search of "give me all the logs I care about for PR X" or "commit S" etc.